phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -945,20 +960,40 @@ public MutationState createIndex(CreateIndexStatement statement, byte[][] splits
-                    PColumn col = resolver.resolveColumn(null, colName.getFamilyName(), colName.getColumnName()).getColumn();
-                    unusedPkColumns.remove(col);
-                    // Ignore view constants for updatable views as we don't need these in the index
-                    if (col.getViewConstant() == null) {
-                        PDataType dataType = IndexUtil.getIndexColumnDataType(col);
-                        colName = ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(col));
-                        allPkColumns.add(new Pair<ColumnName, SortOrder>(colName, pair.getSecond()));
-                        columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), col.isNullable(), col.getMaxLength(), col.getScale(), false, SortOrder.getDefault()));
-                    }
+                for (Pair<ParseNode, SortOrder> pair : indexParseNodeAndSortOrderList) {
+                	ParseNode parseNode = pair.getFirst();
+                	ColumnName colName = null;
+                	// if it is a column 
+                	if (parseNode instanceof ColumnParseNode) {
+                		ColumnParseNode colParseNode = (ColumnParseNode)parseNode;
+	                    PColumn col = resolver.resolveColumn(null, colParseNode.getTableName(), colParseNode.getName()).getColumn();
+	                    unusedPkColumns.remove(col);
+	                    // Ignore view constants for updatable views as we don't need these in the index
+	                    if (col.getViewConstant() != null) 
+	                    	continue;
+	                    colName = ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(col));
+                	}
+            	    // compile the parseNode to get an expression
+                    PhoenixStatement phoenixStatment = new PhoenixStatement(connection);
+                    final StatementContext context = new StatementContext(phoenixStatment, resolver);
+                    ExpressionCompiler expressionCompiler = new ExpressionCompiler(context);
+                    Expression expression = parseNode.accept(expressionCompiler);   
+                    if (expressionCompiler.isAggregate()) {
+                        throw new SQLExceptionInfo.Builder(SQLExceptionCode.AGGREGATE_EXPRESSION_NOT_ALLOWED_IN_INDEX).build().buildException();
+                    }
+                    if (expression.getDeterminism() != Determinism.ALWAYS) {
+                        throw new SQLExceptionInfo.Builder(SQLExceptionCode.NON_DETERMINISTIC_EXPRESSION_NOT_ALLOWED_IN_INDEX).build().buildException();
+                    }
+                	
+                	PDataType dataType = IndexUtil.getIndexColumnDataType(expression.isNullable(), expression.getDataType());
+                	colName = colName != null ? colName : 
+                		ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(null, expression.toString().replaceAll("\"", "")));
```

Comment:
If expression refers to a col that is case sensitive, the string will contain "" which is not a valid character for column names.